libkbfs/block_retrieval_queue.go

```diff
@@ -263,44 +279,69 @@ func (brq *blockRetrievalQueue) CacheAndPrefetch(ctx context.Context,
 	default:
 		// We should return the error here because otherwise we could thrash
 		// the prefetcher.
+		prefetchErrCh <- struct{}{}
 		return err
 	}
 	// This must be called in a goroutine to prevent deadlock in case this
 	// CacheAndPrefetch call was triggered by the prefetcher itself.
 	go func() {
-		<-brq.Prefetcher().PrefetchAfterBlockRetrieved(block,
-			ptr, kmd)
+		prefetcher := brq.Prefetcher()
+		childPrefetchDoneCh, childPrefetchErrCh, numBlocks :=
+			prefetcher.PrefetchAfterBlockRetrieved(block, ptr, kmd)
+
+		// If we have child blocks to prefetch, wait for them.
+		if numBlocks > 0 {
+			for i := 0; i < numBlocks; i++ {
+				select {
+				case <-childPrefetchDoneCh:
+					// Passthrough
+				case <-childPrefetchErrCh:
+					// One error means this block didn't finish prefetching.
+					prefetchErrCh <- struct{}{}
```

Comment:
Definitely a valid concern. It almost seems like the prefetcher needs a cache to track every block currently being prefetched. Still digesting...

Reply:
The idea here is that when `prefetchErrCh` returns, the receiver does nothing to the state of `DonePrefetch` on a block. We need such a mechanism because we don't want to waltz all over other prefetches that are happening. But as you identified in another comment, this can cause havoc when multiple prefetches might need to affect the same block. Essentially, these are the scenarios we need to support:
1. First prefetch for block `c` from parent `a`:
    * A counting semaphore (currently a `chan struct{}`) that reaches 0 when all child block trees are fully prefetched. When this happens, the state of the block is updated to `DonePrefetch = true`.
    * A semaphore that is notified as soon as any child block fails to prefetch. Now it doesn't matter if other blocks finish, so we can short circuit this process, leave `DonePrefetch` alone, and notify the parent's error semaphore.
2. Second prefetch for already-prefetched block `c` from `a'`:
    * Notice that this block already has `DonePrefetch` set, so we send to the parent's `done` semaphore. This is currently incorrect in the code, we are only checking for `triggeredPrefetch`.
3. Second prefetch for triggered but not done for block `c`, originally triggered by `a` and now triggered by `a'`:
    1. Prefetch from `a` is ongoing:
        * We probably need to track this, notice it, and add the `a'`'s done semaphore to a slice of parent channels to notify (this is new code). This also needs to be done atomically, such that we can't race with the completion of the existing prefetch.
    2. Prefetch from `a` failed:
        * Once we check `donePrefetch` instead of `triggeredPrefetch`, this should just work.

Reply:
@strib: I'll work on fixing that scaffolding. Sorry this has been such a mess; it's proven to be quite a challenge to keep the existing prefetch behavior working along the same codepath as the new deep tree syncing.

Reply:
This sounds like a good plan to me.  Thanks for writing it out!  It would probably be helpful as a comment in the code somewhere too.